ni_measurementlink_service/_internal/parameter/_message.py

```diff
@@ -0,0 +1,113 @@
+import struct
+from typing import Any, Dict
+
+from google.protobuf.internal import encoder, wire_format
+from google.protobuf.message import Message
+
+from ni_measurementlink_service._internal.parameter._serializer_types import (
+    Decoder,
+    Encoder,
+    Key,
+    NewDefault,
+)
+
+
+def _message_encoder_constructor(field_index: int, is_repeated: bool, is_packed: bool) -> Encoder:
+    """Mimics google.protobuf.internal.MessageEncoder.
+
+    We cannot use MessageEncoder because it is for the pure-Python implementation
+    of messages. Our messages are upb messages, so we need to call 'SerializeToString'
+
+    See EncodeField:
+    https://github.com/protocolbuffers/protobuf/blob/0b817d46d4ca1977d3dccf2442aeee3c9e98e3a1/python/google/protobuf/internal/encoder.py#L765C15-L765C15
+    """
+    tag = encoder.TagBytes(field_index, wire_format.WIRETYPE_LENGTH_DELIMITED)
+    encode_varint = _varint_encoder()
+
+    def _encode_message(write, value, deterministic):
+        write(tag)
+        bytes = value.SerializeToString()
+        encode_varint(write, len(bytes), deterministic)
+        write(bytes)
+
+    return _encode_message
+
+
+def _varint_encoder():
+    """Return an encoder for a basic varint value (does not include tag).
+
+    From google.protobuf.internal.encoder.py _VarintEncoder
+    """
+    local_int2byte = struct.Struct(">B").pack
+
+    def encode_varint(write, value, unused_deterministic=None):
+        bits = value & 0x7F
+        value >>= 7
+        while value:
+            write(local_int2byte(0x80 | bits))
+            bits = value & 0x7F
+            value >>= 7
+        return write(local_int2byte(bits))
+
+    return encode_varint
+
+
+def _message_decoder_constructor(
+    field_index: int, is_repeated: bool, is_packed: bool, key: Key, new_default: NewDefault
+) -> Decoder:
+    """Mimics google.protobuf.internal.MessageDecoder.
+
+    We cannot use MessageDecoder because it is for the pure-Python implementation
+    of messages. Our messages are upb messages, so we need to call 'ParseFromString'
```

Comment:
```suggestion
    This function was forked in order to call ParseFromString instead of _InternalParse.
    
    _InternalParse is only defined for the pure-Python protobuf implementation. Our child messages
    (like DoubleXY) are defined in .proto files, so they use whichever protobuf implementation that
    google.protobuf.internal.api_implementation chooses (usually upb).
```
```